### PR TITLE
GAWB-969: admin endpoint to rebuild entire ES index

### PR DESCRIPTION
**Pull request by davidangb**
Admin-class endpoint to delete the entire ES index and re-populate it with all published workspaces.

The query to get published workspaces is hugely inefficient, but blocked by GAWB-1059. Once GAWB-1059 lands, this code needs an update.

To see this in action:
1. use the putLibraryMetadata endpoint, via swagger, to create some library metadata on one or more workspaces. Make sure to include "library:published":"true".
2. call the reindexAll endpoint, as an admin.
3. manually query our elasticsearch cluster to see the results:
   http://elasticsearch1.dsde-dev.broadinstitute.org:9200/_cat/indices?v
   http://elasticsearch1.dsde-dev.broadinstitute.org:9200/librarylocal/dataset/_search

(this will all get easier as other stories land)
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  - Acceptance criteria exists and is met
  - Note any changes to implementation from the description
  - Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: build.sbt
Comment (davidangb) [inline build.sbt:75]: elasticsearch uses logging at startup, so I moved the log jars first.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala]: once #257 is merged, the "library:published" value should be centralized.

Also, I left the logging in here as a reminder for us to rewrite this code once GAWB-1059 lands.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: I'm not certain why `sender` works in rawls, but orch requires `context.parent`. But orch _definitely_ needs `context.parent`; messages were getting dropped.

File: src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudServiceActor.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudServiceActor.scala]: I'm currently seeing the log messages for initIndex appearing twice in the logs. I'm researching that ...

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: talked to Doug in person, and he solved it; `sender` is fine, as long as the braces are in the right place.

File: build.sbt
Comment (rushtong) [reply build.sbt:75]: I couldn't find anything useful to explain compilation order. Would love to know why sbt doesn't have all dependencies pulled in before startup.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala]: what does the lead underscore mean?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala]: log timing as well?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala:28]: do you want to execute ES requests in Futures? Doing so would make it match other DAO calls that use the spray client and might be less block-y.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/SearchDAO.scala
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/SearchDAO.scala]: I find it confusing that there is no function signature... is this the same as `def initIndex(): Unit`?

Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/SearchDAO.scala]: it is the same; I can add `: Unit`

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala]: that it's a local/private implementation of a method with the same name (minus the underscore). Totally up for renaming.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala:28]: The ES client has its own async/future handling, based on java.util.concurrent.Future (and heavily subclassed). I did intend to look at it, but was daunted by the possible conflicts with Scala futures ... and then never created a follow-on story or mentioned it in the PR description. I'll make a story.

File: build.sbt
Comment (davidangb) [reply build.sbt:75]: I didn't quite understand it either, but this fixed things - would love to see if other people encounter any issues.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAOSupport.scala]: This doesn't seem to be used anywhere

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (jmthibault79) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala]: I'm not sure about "recreate" as a name because it leaves it alone if it exists, in this case.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala:51]: This can be simply `doc =>` but that is just a style choice

File: src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudApp.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudApp.scala:20]: this is embarrassing. We were starting FireCloudServiceActor twice - once in the above line, and once inside startWebServiceActors below. This is what caused the ElasticSearchDAO.initIndex call to be logged twice - it actually did happen twice.

File: src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudApp.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/FireCloudApp.scala:8]: adding `extends LazyLogging` or `with LazyLogging` to our classes seems easier than manually grabbing a logger via `LoggerFactory.getLogger()`.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala
Comment (dmohs) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: Does that mean this needs to be changed back to sender?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala]: This comment just repeats exactly what the code says fairly clearly.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchDAO.scala]: Ditto.

File: src/main/scala/org/broadinstitute/dsde/firecloud/model/Search.scala
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/firecloud/model/Search.scala]: Be consistent across change with these, either way.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala
Comment (dmohs) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala]: Ignore this. I was not looking at the right page.

**Review by dmohs — APPROVED** [review]
Looks good. A couple of suggestions, but don't hold up this PR. No need to re-review either way.

File: src/main/scala/org/broadinstitute/dsde/firecloud/model/Search.scala
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/model/Search.scala]: yup, I missed one - good catch

**Pull request merged**
